File: gisaid_track_aa_subs.py
#!/home/bill/anaconda3/envs/bio2/bin/python
# -*- coding: utf-8 -*-
"""
gisaid_track_aa_subs.py
    Create a CSV file of AA substitutions in human sequences by date and country from a
    GISAID metafile.

@author: Bill Thompson
@license: GPL 3
@copyright: 2021_04_28
"""
import argparse
import sys
import re
import warnings
import pandas as pd

def GetArgs():
    def ParseArgs(parser):
        class Parser(argparse.ArgumentParser):
            def error(self, message):
                sys.stderr.write('error: %s\n' % message)
                self.print_help()
                sys.exit(2)
                
        parser = Parser(description='Reformat GISAID metefile.')
        parser.add_argument('-i', '--input_file',
                            required = True,
                            help = 'GISAID metafile (required).',
                            type = str)
        parser.add_argument('-o', '--output_file',
                            required = True,
                            help = 'Output CSV file.',
                            type = str)
        parser.add_argument('-s', '--host',
                            required = False,
                            help = 'Virus host species (default = Human).',
                            default = "Human",
                            type = str)

        return parser.parse_args()

    parser = argparse.ArgumentParser()
    args = ParseArgs(parser)
    
    return args

def main():
    args = GetArgs()
    meta_file = args.input_file
    out_file = args.output_file
    host = args.host
    
    df = pd.read_csv(meta_file,
                     sep = '\t', 
                     dtype={'Location': str, "Variant": str, "Is reference?": str})
    
    # select columns and filter the data
    df2 = df.filter(["Virus name", "Accession ID", 
                     "Collection date", "Location",
                     "Pango lineage", "AA Substitutions",
                     "Host"])

    # country is the second part of location                      
    df2["Country"] = df2.Location.str.split("/", expand = True)[1] 

    df2 = df2[df2["Host"] == host]   # filter host and remove NAs from lineage
    df2 = df2.dropna(subset = ["Pango lineage"])

    df2 = df2.drop(["Location", "Host"], axis = 1) # We don't need these anymore

    # throw out malformed dates
    # regex for dates
    p = re.compile(r"^\d\d\d\d\-\d\d\-\d\d$")
    df2 = df2[df2["Collection date"].apply(lambda x: True if p.match(x) else False)]

    # clean up Country and AAs
    # these throw warnings, but they are actually OK.
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        df2["Country"] = df2["Country"].str.strip()
        df2["AA Substitutions"] = df2["AA Substitutions"].str.strip("()")
    
    # Explode AA substitutions. Each AA has it's own row
    # from https://stackoverflow.com/questions/50731229/split-cell-into-multiple-rows-in-pandas-dataframe
    df2 = df2.set_index(["Virus name", "Accession ID", 
                     "Collection date", "Country",
                     "Pango lineage"]).apply(lambda x: x.str.split(',').explode()).reset_index()
    
    df2.to_csv(out_file, index = False)

if __name__ == '__main__':
    main()

File: gisaid_track_aa_subs.jl
#=
gisaid_track_aa_subs.jl
    Create a CSV file of AA substituions in human sequences by date and country from a
    GISAID metafile.

@author
    Bill Thompson
@copyright
    GPL 3
@date
    2021_04-25
=#

using DataFrames
using CSV
using Dates
using ArgParse

function GetArgs()
    # read arguments from command line
    # from https://argparsejl.readthedocs.io/en/latest/argparse.html
    s = ArgParseSettings(description = "Extract AA substituions from GISAID metefile.",
                         version = "1.0",
                         add_version = true)

    @add_arg_table s begin
        "--input_file", "-i"
        required = true
        help = "GISAID meteafile, (required)."
        arg_type = String

        "--output_file", "-o"
        required = true
        help = "CSV file with results. (required)."
        arg_type = String

        "--host", "-s"
        required = false
        help = "Host species (default = Human)."
        arg_type = String
        default = "Human"
    end

    return parse_args(s)
end

function check_date(date_str)
    #=
    Check that data is formatted as yyyy-mm-dd.

    @arguments
    data_str - String
        A string representing a date.

    @returns
    A Date in yyyy-mm-dd of nothing if string is not properly formatted.
    =#
    r = r"^\d\d\d\d\-\d\d\-\d\d$"
    if isnothing(match(r, date_str))
        return nothing
    end

    return Date(date_str, "yyyy-mm-dd")
end

function main()
    #=
    Check sequences for valid date and host. 
    Write a CSV file with a row for each AA substitution.
    =#
    args = GetArgs()
    meta_table = args["input_file"]
    aa_subs_table = args["output_file"]
    host = args["host"]
   
    # Read the metfile and then make a new empty dataframe.
    df = DataFrame(CSV.File(meta_table, delim = '\t'))
    df2 = DataFrame(Date = Date[], 
                    Country = String[],
                    Virus_name = String[], 
                    Accession_ID = String[], 
                    Pango_lineage = String[],
                    AA_subs = String[])

    for i in 1:size(df)[1]
        if i % 10000 == 0
            println(i)
        end
        
        if df[i, "Host"] != host
            continue
        end

        if ismissing(df[i, "Pango lineage"])
            continue
        end

        date = check_date(df[i, "Collection date"])
        if isnothing(date)
            continue
        end
        
        aas = split(df[i, "AA Substitutions"][2:end-1], ",")
        if isempty(aas)
            continue
        end

        country = strip(split(df[i, "Location"], "/")[2], ' ')

        for aa in aas
            # Debugging - just in case.
            # println(i, [date,
            #           country,
            #           df[i, "Virus name"],
            #           df[i, "Accession ID"],
            #           df[i, "Pango lineage"],
            #           aa])
            push!(df2, [date,
                        country, 
                        df[i, "Virus name"],
                        df[i, "Accession ID"],
                        df[i, "Pango lineage"],
                        aa])
        end
    end

    CSV.write(aa_subs_table, df2)
end

main()